File: leetcode/trie/explore/practical_application/replace_words.cpp
#include <vector>
#include <string>
#include <sstream>

using namespace std;

class Solution {
public:
    string replaceWords(vector<string>& dict, string sentence) {
        istringstream ss(sentence);
        string result;
        result.reserve(sentence.length());
        do {
            string word;
            ss >> word;

            for (auto root : dict)
            {
                if (word.find(root) == 0)
                {
                    word = root;
                    break;
                }
            }
            result += word;
            result += " ";
        } while (ss);
        result.pop_back();
        result.pop_back();

        return result;
    }
};

int main() {
    Solution example;
    vector<string> dict = {
        "cat", "bat", "rat"
    };
    string sentence = "the cattle was rattled by the battery";

    example.replaceWords(dict, sentence);
}

File: leetcode/trie/explore/practical_application/replace_words.py
class Solution:
    def replaceWords(self, dict: List[str], sentence: str) -> str:
        result = ""
        sentence = sentence.split(' ')
        for word in sentence:
            for root in dict:
                if word.find(root) == 0:
                    word = root
                    break
            result += word + " "

        return result[:-1]